### PR TITLE
custom indicator added

### PR DESCRIPTION
**Pull request by hamxa152**
**Ticket No:**  #203 
**YMAL Code:** 
      - Text:
          id: status
          text: Carousel's initial selected index is ${carousel.selectedIndex}

      - Carousel:
          id: carousel
          styles: 
            height: 100
            gap: 20
            layout: auto
            borderColor: blue
            padding: 10
            borderWidth: 2
            indicatorWidget: MyIndicator
            selectedIndicatorWidget: MySelectedIndicator
            # indicatorType: rectangle
            indicatorPosition: top
            indicatorWidth: 30
            indicatorHeight: 2
          onItemChange: |
            //@code
            status.text = 'Selected index is now ' + this.selectedIndex;

MyIndicator:
  Icon:
    icon: star
    styles: { size: 12 }

MySelectedIndicator:
  Icon:
    icon: star
    styles: { color: 0xFFFF0000 }

### CODE REVIEW

**Comment by vusters** [comment]
@hamxa152 valid YAML indentation please.

File: lib/widget/carousel.dart [hunk 1]
```diff
@@ -150,7 +157,7 @@ class CarouselState extends WidgetState<Carousel> with TemplatedWidgetState {
       List<Widget> indicators = [];
       for (int i = 0; i < items.length; i++) {
         indicators.add(GestureDetector(
-          child: getIndicator(i == focusIndex),
+          child: getIndicator(i == focusIndex , scopeManager!),
```
Comment (vusters) [inline lib/widget/carousel.dart:160]: please don't do this.

File: lib/widget/carousel.dart [hunk 2]
```diff
@@ -166,7 +173,7 @@ class CarouselState extends WidgetState<Carousel> with TemplatedWidgetState {
       // Carousel requires a fixed height, so to make sure the indicators don't shift the UI, we'll make
       // sure there's at least 1 invisible indicator that takes up the space
       if (indicators.isEmpty) {
-        indicators.add(Opacity(child: getIndicator(false), opacity: 0));
+        indicators.add(Opacity(child: getIndicator(false , scopeManager!), opacity: 0));
```
Comment (vusters) [inline lib/widget/carousel.dart:176]: please don't do this.

File: lib/widget/carousel.dart [hunk 3]
```diff
@@ -153,11 +159,11 @@ class CarouselState extends WidgetState<Carousel> with TemplatedWidgetState {
     );
 
     // show indicators
-    if (widget._controller.indicatorType != null && widget._controller.indicatorType != IndicatorType.none) {
+    if (widget._controller.indicatorType != null && widget._controller.indicatorType != IndicatorType.none || customIndicator != null || selectedCustomIndicator != null) {
```
Comment (vusters) [inline lib/widget/carousel.dart:162]: don't mix && and || without parenthesis. Be explicit about your intention, as different languages evaluate them differently.

File: lib/widget/carousel.dart [hunk 4]
```diff
@@ -142,6 +145,9 @@ class CarouselState extends WidgetState<Carousel> with TemplatedWidgetState {
   @override
   Widget buildWidget(BuildContext context) {
     ScopeManager? scopeManager = DataScopeWidget.getScope(context);
+
+    customIndicator = defineIndicatorWidget(scopeManager!);
```
Comment (vusters) [inline lib/widget/carousel.dart:149]: You have to understand why we don't want to do `!` - don't just move code around. Please spend a few minutes to understand why.

**Review by vusters — DISMISSED** [review]

**Comment by vusters** [comment]
@vinothvino42 please close this PR once you resolve this issue

File: lib/widget/carousel.dart (hunk 2 above)
Comment (vinothvino42) [reply lib/widget/carousel.dart]: Fixed by @hamxa152

File: lib/widget/carousel.dart (hunk 1 above)
Comment (vinothvino42) [reply lib/widget/carousel.dart]: @vusters It is now fixed by @hamxa152

File: lib/widget/carousel.dart (hunk 3 above)
Comment (vinothvino42) [reply lib/widget/carousel.dart:162]: I updated it like this

```dart
bool isCustomIndicator =
        customIndicator != null || selectedCustomIndicator != null;
    if (widget._controller.indicatorType != null &&
            widget._controller.indicatorType != IndicatorType.none ||
        isCustomIndicator) {
```

**Review by vinothvino42 — APPROVED** [review]

File: lib/widget/carousel.dart [hunk 5]
```diff
@@ -288,7 +299,36 @@ class CarouselState extends WidgetState<Carousel> with TemplatedWidgetState {
     );
   }
 
+  Widget? _buildIndicatorWidget(ScopeManager? scopeManager) {
```
Comment (snehmehta) [inline lib/widget/carousel.dart:302]: Can we have only one buildWidget?

Comment (vinothvino42) [reply lib/widget/carousel.dart:302]: Do you mean to have one build method for creating the indicator? @snehmehta 
Okay, I'll combine both into a single method.

Comment (vinothvino42) [reply lib/widget/carousel.dart]: @snehmehta I updated the code and can you check the `_buildIndicatorWidget` method?

**Review by snehmehta — APPROVED** [review]

**Pull request merged**